Clean up request context serialization

gavin-norman-sociomantic opened the pull request:
Work in progress.

david-eckardt-sociomantic commented on src/swarm/neo/util/StructSerializer.d:
```diff
@@ -0,0 +1,335 @@
+    {
+        static if ( isDynamicArrayType!(typeof(f)) )
+            serializeDynamicArray(f, buf);
+    }
```
Alternative, slightly shorter:
```d
// Do a simple, binary copy of the struct.
buf.length = S.sizeof;
enableStomping(buf);
auto serialized_struct = cast(S*)buf.ptr;
*serialized_struct = s;
// Handle any array fields.
foreach ( i, ref f; (*serialized_struct).tupleof )
// ...
```

gavin-norman-sociomantic replied:
Ah yeah, good idea.

gavin-norman-sociomantic commented:
Gah damn, I just noticed that this restriction: https://github.com/sociomantic-tsunami/swarm/pull/178/files#diff-4195e2bb4e0d50489a9738a2c43711a1R199 makes this serializer unsuitable for the task it's intended for. I was thinking we're serializing the args struct containing the arguments provided by the user. That struct is actually nested inside two others, though: https://github.com/sociomantic-tsunami/swarm/blob/v4.x.x/src/swarm/neo/client/mixins/RequestCore.d#L179 and https://github.com/sociomantic-tsunami/swarm/blob/v4.x.x/src/swarm/neo/client/mixins/RequestCore.d#L249.

Might need to rethink this...

gavin-norman-sociomantic commented:
Updated with support for arrays inside nested structs.

Next step: adapt internal request context serialization to use this.

gavin-norman-sociomantic commented:
Updated: fixed D2 errors.

david-eckardt-sociomantic commented on src/swarm/neo/util/StructSerializer.d:
```diff
@@ -0,0 +1,347 @@
+private void serializeDynamicArray ( A ) ( ref A array, ref void[] buf )
+{
+    static assert(isDynamicArrayType!(A));
+    static assert(isPrimitiveType!(BaseTypeOfArrays!(A)));
```
Suggest
- using `A` as the array *element* type (unless is causes some type qualifier related complication) and
- disallowing indirections other than arrays,  i.e. also allowing struct/union/static array elements as long as they are plain values, instead of allowing elements of primitive type only.
```d
private void serializeDynamicArray ( A ) ( ref A[] array, ref void[] buf )
{
    // ...

    // Recursively serialize arrays of arrays.
    static if (isDynamicArrayType!(A))
    {
        foreach ( ref e; array )
            serializeDynamicArray(e, buf);
    }
    else
        static assert(!hasIndirections!(A));
}
```

gavin-norman-sociomantic commented:
Updated: adapt swarm client's internals to use the new serializer.

This adaptation contains at least one breaking change, so I'll have to consider whether to try to do as much as I can without the breaking change or whether to just move this PR to v5.0.0.

mihails-strasuns-sociomantic commented on src/swarm/neo/util/StructSerializer.d:
```diff
@@ -0,0 +1,349 @@
+
+    // Test helper function to serialize, deserialize, and compare against the
+    // original.
+    bool serialzeDeserialize ( S ) ( S s )
```
typoe `serialze`

david-eckardt-sociomantic commented on src/swarm/neo/util/StructSerializer.d:
```diff
@@ -0,0 +1,349 @@
+
+*******************************************************************************/
+
+public S* deserialize ( S ) ( Const!(void)[] serialized )
```
This casts the `const` away from `serialized`. It looks like a case for `inout`,
```d
Inout!(S)* deserialize ( S ) ( Inout!(void)[] serialized )
```
so that the qualifier of the return type follows that of the argument.

gavin-norman-sociomantic replied:
Mm I don't see why that would be useful...?

1. `Const!(void)[] serialized` states that we never modify the input buffer.
2. `S*` states that we return a pointer to some type. (Note that the user can do `auto s = deserialize!(Const!(Whatever))(buffer);` if they want to.)

david-eckardt-sociomantic commented on src/swarm/neo/util/StructSerializer.d:
```diff
@@ -0,0 +1,349 @@
+
+*******************************************************************************/
+
+private void seriaizeStructDynamicArrays ( S ) ( ref S s, ref void[] buf )
```
Typo `seriaize`

gavin-norman-sociomantic commented:
Updated:
* `deserialize` function now accepts a mutable buffer. This is perfectly fine for our usage (we always deserialize from a mutable buffer) and prevents the compiler from making any assumptions about the buffer potentially being immutable.

Commit:
Adapt request context serialization to use new serializer

gavin-norman-sociomantic commented:
Updated:
* Renamed to un/packer, as suggested by Mihails.
* Simplified `serializeDynamicArray`, as suggested by David.

mihails-strasuns-sociomantic commented on src/swarm/neo/util/StructPacker.d:
```diff
@@ -0,0 +1,349 @@
+/*******************************************************************************
+
+    Functions to pack a struct instance and all contained array buffers into a
+    contiguous chunk of memory.
```
I would explicitly stress that packed data can't be copied/moved because it contains interior pointers.

gavin-norman-sociomantic replied:
Mentioned in method doc.

mihails-strasuns-sociomantic replied:
Better to also put it into the module doc as it is the main design point of the whole concept, immediately answering "how is it different from serializer" question.

gavin-norman-sociomantic replied:
:+1: That's a minor difference, I'd say. I'll add something, anyway.

gavin-norman-sociomantic replied:
Updated.

Commit:
Directly serialize const context

Using the new packer functions, there is no longer any need to serialize
the user-provided args into an intermediary buffer first.

gavin-norman-sociomantic commented:
Updated: replaced breaking change with deprecation path.

gavin-norman-sociomantic commented:
Updated:
* Tested with dhtproto.
* Added comment as suggested by Mihails.
* Unblocked.

gavin-norman-sociomantic commented:
Ah... release notes coming up.

Commit:
Store request resources object directly as object, not as ubyte[]

The new serializer (used to serialize the request context in the Request
instance) can handle reference types, so we no longer need to use the
SerializableReferenceType workaround to store the request resources.

gavin-norman-sociomantic commented:
Updated with release notes for the deprecation. I decided not to add release notes for the new feature (the un/packer functions), as they're purely internal to the client and not intended to be used externally.

nemanja-boric-sociomantic commented on src/swarm/neo/util/StructPacker.d:
```diff
@@ -0,0 +1,354 @@
+    contiguous chunk of memory.
+
+    Copyright:
+        Copyright (c) 2016-2017 sociomantic labs GmbH. All rights reserved
```
Check copyright (only 2017 needed).

gavin-norman-sociomantic replied:
Updated.

Commit:
Add functions to pack a struct + arrays into a contiguous chunk of memory

Fixes #177.

gavin-norman-sociomantic commented:
Any further comments on this?

david-eckardt-sociomantic commented on src/swarm/neo/client/mixins/RequestCore.d:
```diff
@@ -259,11 +259,42 @@ public template RequestCore ( RequestType request_type_, ubyte request_code,
+
+        private struct RequestResources
+        {
+            private Object* request_resources;
```
Why is this a pointer?

gavin-norman-sociomantic replied:
So that the return value of `get` always reflects the value of `shared_resources` in the outer context.

gavin-norman-sociomantic replied:
It's a bit theoretical, but you could do:
```D
auto rr = context.request_resources();
do_something(rr.get());
context.shared_resources = something_else;
do_something(rr.get()); // should reflect the changed value
```

david-eckardt-sociomantic replied:
I see. Fair enough.

david-eckardt-sociomantic commented on src/swarm/neo/util/StructPacker.d:
```diff
@@ -0,0 +1,360 @@
+        {
+            // ...they only contain primitive types.
+            static if ( !isPrimitiveType!(BaseTypeOfArrays!(F)) )
+                error = true;
```
- For the sake of an error message that is more comprehensive than "something is wrong" I suggest using `static assert`s on the error conditions with the unsupported type and maybe even the field name in the message, and naming the function `void checkPackable` or something similar.
- Why are dynamic arrays of pointers/delegates/objects etc. disallowed, again, while static arrays of such types are allowed?

gavin-norman-sociomantic replied:
>Why are dynamic arrays of pointers/delegates/objects etc. disallowed, again, while static arrays of such types are allowed?

Because dynamic array contents are copied to the end of the packed buffer and may thus become misaligned.

david-eckardt-sociomantic replied:
Ah, right. Now I remember.

gavin-norman-sociomantic replied:
I'll add a comment explaining this.

david-eckardt-sociomantic commented on src/swarm/neo/util/StructPacker.d:
```diff
@@ -0,0 +1,360 @@
+            // ...they could be packed on their own.
+            static if ( !canPack!(F) )
+                error = true;
+        }
```
Not sure if it is that important, but you may want to check unions and allow only those that don't contain dynamic arrays.

gavin-norman-sociomantic commented:
Updated.
